File: src/main/java/org/woahoverflow/po/Discord.java
package org.woahoverflow.po;

import net.arikia.dev.drpc.DiscordEventHandlers;
import net.arikia.dev.drpc.DiscordRPC;
import net.arikia.dev.drpc.DiscordRichPresence;

import java.util.concurrent.ConcurrentHashMap;

import static org.woahoverflow.po.PresenceOverflow.ui;

public final class Discord
{
    private static String state = "";
    private static String details = "";
    private static String bigImageCaption = "";
    private static String bigImage = "";
    private static String smallImageCaption = "";
    private static String smallImage = "";
    private static final DiscordEventHandlers HANDLERS = new DiscordEventHandlers.Builder().setReadyEventHandler((user) ->
    {
        ui.updateStatus("Running... " + user.username);
        DiscordRPC.discordUpdatePresence(build());
    }).build();

    public static void initDiscord()
    {
        DiscordRPC.discordInitialize(ui.getValues().get("app_id"), HANDLERS, true);
        ui.updateStatus("Running...");
        DiscordRPC.discordRunCallbacks();
    }

    private static void update()
    {
        ConcurrentHashMap<String, String> hash = ui.getValues();
        state = hash.get("state");
        details = hash.get("details");
        bigImageCaption = hash.get("bigImageCaption");
        smallImageCaption = hash.get("smallImageCaption");
        bigImage = hash.get("big_image_key");
        smallImage = hash.get("small_image_key");

    }

    public static void refresh()
    {
        DiscordRPC.discordUpdatePresence(build());
    }
    private static DiscordRichPresence build()
    {
        update();
        if (state.isEmpty())
        {
            ui.updateStatus("Invalid State!");
            return null;
        }

        DiscordRichPresence.Builder rp = new DiscordRichPresence.Builder(state);

        if (!bigImage.isEmpty()) {
            rp.setBigImage(bigImage, bigImageCaption);
        }

        if (!smallImage.isEmpty()) {
            rp.setSmallImage(smallImage, smallImageCaption);
        }

        if (!details.isEmpty()) {
            rp.setDetails(details);
        }

        return rp.build();
    }
}

File: presenceoverflow/src/main/java/org/woahoverflow/po/Discord.kt
package org.woahoverflow.po

import net.arikia.dev.drpc.DiscordEventHandlers
import net.arikia.dev.drpc.DiscordRPC
import net.arikia.dev.drpc.DiscordRichPresence
import org.woahoverflow.po.ui.panel
import java.awt.Color

private var handlers: DiscordEventHandlers? = null
private var running = false
private var previousId: String? = null

fun init() {
    PresenceOverflow.LOGGER.debug("Running init for Discord")
    handlers = DiscordEventHandlers.Builder().setReadyEventHandler {
        PresenceOverflow.LOGGER.debug("User {} is now connected.", it.username)
        panel.statusLabel.text = "User connected"
        panel.statusLabel.foreground = Color.GREEN
    }.setDisconnectedEventHandler { _, st ->
        panel.statusLabel.text = "User disconnected"
        panel.statusLabel.foreground = Color.RED
        PresenceOverflow.LOGGER.debug("User has disconnected. \n{}", st)
    }.build()
}

fun refresh(id: String, profile: Profile) {
    panel.statusLabel.text = "Loading..."
    panel.statusLabel.foreground = Color.YELLOW

    if (handlers == null) init()

    if (previousId != id) {
        if (running) DiscordRPC.discordShutdown()

        PresenceOverflow.LOGGER.debug("Running init for DiscordRPC")
        DiscordRPC.discordInitialize(id, handlers, true)
    }

    if (profile.state == "") {
        PresenceOverflow.LOGGER.debug("The state is empty, returning.")
        panel.statusLabel.text = "Invalid State!"
        panel.statusLabel.foreground = Color.RED
        return
    }

    val presence = DiscordRichPresence.Builder(profile.state)

    if (profile.details != "") presence.setDetails(profile.details)
    if (profile.bigImageKey != "") presence.setBigImage(profile.bigImageKey, profile.bigImageCaption)
    if (profile.smallImageKey != "") presence.setSmallImage(profile.smallImageKey, profile.smallImageCaption)

    DiscordRPC.discordUpdatePresence(presence.build())
    panel.statusLabel.text = "Updated!"
    panel.statusLabel.foreground = Color.GREEN
}